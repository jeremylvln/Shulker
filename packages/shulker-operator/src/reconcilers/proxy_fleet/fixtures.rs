use std::collections::BTreeMap;

use google_agones_crds::v1::fleet_autoscaler::{
    FleetAutoscalerPolicy, FleetAutoscalerPolicyBufferSpec, FleetAutoscalerPolicySpec,
};
use http::{Request, Response};
use k8s_openapi::{api::core::v1::EnvVar, apimachinery::pkg::util::intstr::IntOrString};
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::{
    resourceref::ResourceRefSpec,
    schemas::{FleetAutoscalingSpec, TemplateSpec},
    v1alpha1::{
        minecraft_cluster::MinecraftClusterRef,
        proxy_fleet::{
            ProxyFleet, ProxyFleetServiceExternalTrafficPolicy, ProxyFleetServiceSpec,
            ProxyFleetServiceType, ProxyFleetSpec, ProxyFleetTemplateConfigurationSpec,
            ProxyFleetTemplatePodOverridesSpec, ProxyFleetTemplateSpec, ProxyFleetTemplateVersion,
            ProxyFleetTemplateVersionSpec,
        },
    },
};

lazy_static! {
    pub static ref TEST_PROXY_FLEET: ProxyFleet = ProxyFleet {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-proxy".to_string()),
            ..ObjectMeta::default()
        },
        spec: ProxyFleetSpec {
            cluster_ref: MinecraftClusterRef::new("my-cluster".to_string()),
            replicas: 3,
            template: TemplateSpec {
                metadata: Some(ObjectMeta {
                    labels: Some(BTreeMap::from([(
                        "test-label/shulkermc.io".to_string(),
                        "my-value".to_string()
                    )])),
                    annotations: Some(BTreeMap::from([(
                        "test-annotation/shulkermc.io".to_string(),
                        "my-value".to_string()
                    )])),
                    ..ObjectMeta::default()
                }),
                spec: ProxyFleetTemplateSpec {
                    version: ProxyFleetTemplateVersionSpec {
                        channel: ProxyFleetTemplateVersion::Velocity,
                        name: "latest".to_string(),
                    },
                    config: ProxyFleetTemplateConfigurationSpec {
                        existing_config_map_name: None,
                        plugins: Some(vec![ResourceRefSpec {
                            url: Some("https://example.com/my_plugin.jar".to_string()),
                            ..ResourceRefSpec::default()
                        }]),
                        patches: Some(vec![ResourceRefSpec {
                            url: Some("https://example.com/my_patch.tar.gz".to_string()),
                            ..ResourceRefSpec::default()
                        }]),
                        max_players: 1000,
                        motd: "A Motd".to_string(),
                        server_icon: "abc==".to_string(),
                        proxy_protocol: true,
                        ttl_seconds: 3600,
                        players_delta_before_exclusion: 15,
                        skip_agent_download: false
                    },
                    pod_overrides: Some(ProxyFleetTemplatePodOverridesSpec {
                        image: None,
                        env: Some(vec![EnvVar {
                            name: "EXTRA_ENV".to_string(),
                            value: Some("my_value".to_string()),
                            ..EnvVar::default()
                        }]),
                        resources: None,
                        affinity: None,
                        node_selector: Some(BTreeMap::from([(
                            "beta.kubernetes.io/os".to_string(),
                            "linux".to_string()
                        )])),
                        tolerations: None,
                        service_account_name: None,
                        volume_mounts: None,
                        volumes: None,
                        ports: None,
                    })
                },
            },
            service: Some(ProxyFleetServiceSpec {
                type_: ProxyFleetServiceType::LoadBalancer,
                annotations: Some(BTreeMap::from([(
                    "service.beta.kubernetes.io/load-balancer".to_string(),
                    "internal".to_string()
                )])),
                external_traffic_policy: Some(ProxyFleetServiceExternalTrafficPolicy::Cluster),
                preferred_reconnection_address: Some("127.0.0.1".to_string())
            }),
            autoscaling: Some(FleetAutoscalingSpec {
                agones_policy: Some(FleetAutoscalerPolicySpec {
                    type_: FleetAutoscalerPolicy::Buffer,
                    buffer: Some(FleetAutoscalerPolicyBufferSpec {
                        max_replicas: 8,
                        min_replicas: 3,
                        buffer_size: IntOrString::Int(2),
                    }),
                    webhook: None,
                })
            }),
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
