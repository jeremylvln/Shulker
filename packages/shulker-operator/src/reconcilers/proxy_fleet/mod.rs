use std::{collections::BTreeMap, sync::Arc, time::Duration};

use futures::StreamExt;
use google_agones_crds::v1::{fleet::Fleet, fleet_autoscaler::FleetAutoscaler};
use k8s_openapi::api::core::v1::{ConfigMap, Service};
use kube::{
    api::{ListParams, PatchParams},
    runtime::{controller::Action, watcher::Config, Controller},
    Api, Client, ResourceExt,
};
use tracing::*;

use shulker_crds::{
    condition::{ConditionStatus, HasConditions},
    v1alpha1::proxy_fleet::{ProxyFleet, ProxyFleetStatus},
};

use self::{
    config_map::ConfigMapBuilder, fleet::FleetBuilder, fleet_autoscaler::FleetAutoscalerBuilder,
    service::ServiceBuilder,
};

use super::{builder::reconcile_builder, status::patch_status, ReconcilerError, Result};

mod config_map;
mod fleet;
mod fleet_autoscaler;
mod service;

#[cfg(test)]
mod fixtures;

struct ProxyFleetReconciler {
    client: kube::Client,

    // Builders
    config_map_builder: ConfigMapBuilder,
    service_builder: ServiceBuilder,
    fleet_builder: FleetBuilder,
    fleet_autoscaler_builder: FleetAutoscalerBuilder,
}

impl ProxyFleetReconciler {
    async fn reconcile(
        &self,
        api: Api<ProxyFleet>,
        proxy_fleet: Arc<ProxyFleet>,
    ) -> Result<Action> {
        reconcile_builder(&self.config_map_builder, proxy_fleet.as_ref()).await?;
        reconcile_builder(&self.service_builder, proxy_fleet.as_ref()).await?;
        let fleet = reconcile_builder(&self.fleet_builder, proxy_fleet.as_ref()).await?;
        reconcile_builder(&self.fleet_autoscaler_builder, proxy_fleet.as_ref()).await?;

        if let Some(fleet) = &fleet {
            if let Some(fleet_status) = &fleet.status {
                let mut proxy_fleet = proxy_fleet.as_ref().clone();
                if proxy_fleet.status.is_none() {
                    proxy_fleet.status = Some(ProxyFleetStatus::default());
                }

                let status = proxy_fleet.status.as_mut().unwrap();

                status.replicas = fleet_status.replicas;
                status.ready_replicas = fleet_status.ready_replicas;
                status.allocated_replicas = fleet_status.allocated_replicas;

                if status.ready_replicas > 0 || status.allocated_replicas > 0 {
                    status.set_condition(
                        "Available".to_string(),
                        ConditionStatus::True,
                        "AtLeastOneReadyOrAllocated".to_string(),
                        "One or more servers are ready or allocated".to_string(),
                    );
                } else {
                    status.set_condition(
                        "Available".to_string(),
                        ConditionStatus::False,
                        "NoneReady".to_string(),
                        "No server are ready".to_string(),
                    );
                };

                patch_status(
                    &api,
                    &PatchParams::apply("shulker-operator").force(),
                    &proxy_fleet,
                )
                .await?;
            }
        }

        Ok(Action::requeue(Duration::from_secs(5 * 60)))
    }

    async fn cleanup(&self, proxy_fleet: Arc<ProxyFleet>) -> Result<Action> {
        info!(
            name = proxy_fleet.name_any(),
            namespace = proxy_fleet.namespace(),
            "cleaning up ProxyFleet",
        );

        Ok(Action::await_change())
    }

    fn get_labels(
        proxy_fleet: &ProxyFleet,
        name: String,
        component: String,
    ) -> BTreeMap<String, String> {
        BTreeMap::from([
            ("app.kubernetes.io/name".to_string(), name.clone()),
            (
                "app.kubernetes.io/instance".to_string(),
                format!("{}-{}", name, proxy_fleet.name_any()),
            ),
            ("app.kubernetes.io/component".to_string(), component),
            (
                "app.kubernetes.io/part-of".to_string(),
                format!("cluster-{}", proxy_fleet.spec.cluster_ref.name.clone()),
            ),
            (
                "app.kubernetes.io/managed-by".to_string(),
                "shulker-operator".to_string(),
            ),
            (
                "minecraftcluster.shulkermc.io/name".to_string(),
                proxy_fleet.spec.cluster_ref.name.clone(),
            ),
            (
                "proxyfleet.shulkermc.io/name".to_string(),
                proxy_fleet.name_any(),
            ),
        ])
    }
}

#[instrument(skip(ctx, proxy_fleet))]
async fn reconcile(proxy_fleet: Arc<ProxyFleet>, ctx: Arc<ProxyFleetReconciler>) -> Result<Action> {
    let ns = proxy_fleet.namespace().unwrap();
    let proxy_fleets_api: Api<ProxyFleet> = Api::namespaced(ctx.client.clone(), &ns);

    info!(
        name = proxy_fleet.name_any(),
        namespace = ns,
        "reconciling ProxyFleet",
    );

    if proxy_fleet.metadata.deletion_timestamp.is_none() {
        ctx.reconcile(proxy_fleets_api.clone(), proxy_fleet.clone())
            .await
    } else {
        ctx.cleanup(proxy_fleet.clone()).await
    }
}

fn error_policy(
    _proxy_fleet: Arc<ProxyFleet>,
    error: &ReconcilerError,
    _ctx: Arc<ProxyFleetReconciler>,
) -> Action {
    warn!("reconcile failed: {:?}", error);
    // ctx.metrics.reconcile_failure(&proxy_fleet, error);
    Action::requeue(Duration::from_secs(5))
}

pub async fn run(client: Client) {
    let proxy_fleets_api = Api::<ProxyFleet>::all(client.clone());
    if let Err(e) = proxy_fleets_api.list(&ListParams::default().limit(1)).await {
        error!("CRD is not queryable; {e:?}. Is the CRD installed?");
        std::process::exit(1);
    }

    let context = ProxyFleetReconciler {
        client: client.clone(),
        config_map_builder: ConfigMapBuilder::new(client.clone()),
        service_builder: ServiceBuilder::new(client.clone()),
        fleet_builder: FleetBuilder::new(client.clone()),
        fleet_autoscaler_builder: FleetAutoscalerBuilder::new(client.clone()),
    };

    Controller::new(proxy_fleets_api, Config::default().any_semantic())
        .owns(
            Api::<ConfigMap>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<Service>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<Fleet>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<FleetAutoscaler>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .shutdown_on_signal()
        .run(reconcile, error_policy, context.into())
        .filter_map(|x| async move { std::result::Result::ok(x) })
        .for_each(|_| futures::future::ready(()))
        .await;
}
