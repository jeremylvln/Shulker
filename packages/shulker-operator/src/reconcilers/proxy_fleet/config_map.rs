use std::collections::BTreeMap;

use k8s_openapi::api::core::v1::ConfigMap;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;
use shulker_crds::v1alpha1::proxy_fleet::ProxyFleetTemplateVersion;

use shulker_crds::v1alpha1::proxy_fleet::ProxyFleet;
use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

use super::ProxyFleetReconciler;

pub struct ConfigMapBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl<'a> ResourceBuilder<'a> for ConfigMapBuilder {
    type OwnerType = ProxyFleet;
    type ResourceType = ConfigMap;
    type Context = ();

    fn name(proxy_fleet: &Self::OwnerType) -> String {
        format!("{}-config", proxy_fleet.name_any())
    }

    fn api(&self, proxy_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            proxy_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, proxy_fleet: &Self::OwnerType) -> bool {
        proxy_fleet
            .spec
            .template
            .spec
            .config
            .existing_config_map_name
            .is_none()
    }

    async fn build(
        &self,
        proxy_fleet: &Self::OwnerType,
        name: &str,
        _existing_config_map: Option<&Self::ResourceType>,
        _context: Option<Self::Context>,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let mut data = BTreeMap::from([
            (
                "init-fs.sh".to_string(),
                include_str!("../../../assets/proxy-init-fs.sh").to_string(),
            ),
            (
                "probe-readiness.sh".to_string(),
                include_str!("../../../assets/proxy-probe-readiness.sh").to_string(),
            ),
            (
                "server-icon.png".to_string(),
                proxy_fleet.spec.template.spec.config.server_icon.clone(),
            ),
        ]);

        match proxy_fleet.spec.template.spec.version.channel {
            ProxyFleetTemplateVersion::Velocity => {
                data.insert(
                    "velocity-config.toml".to_string(),
                    velocity::VelocityToml::from_spec(
                        &proxy_fleet.spec.template.spec.config,
                        proxy_fleet.spec.service.as_ref(),
                    )
                    .to_string(),
                );
            }
            _ => {
                data.insert(
                    "bungeecord-config.yml".to_string(),
                    bungeecord::BungeeCordYml::from_spec(
                        &proxy_fleet.spec.template.spec.config,
                        proxy_fleet.spec.service.as_ref(),
                    )
                    .to_string(),
                );
            }
        }

        let config_map = ConfigMap {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(proxy_fleet.namespace().unwrap().clone()),
                labels: Some(ProxyFleetReconciler::get_labels(
                    proxy_fleet,
                    "config".to_string(),
                    "proxy".to_string(),
                )),
                ..ObjectMeta::default()
            },
            data: Some(data),
            ..ConfigMap::default()
        };

        Ok(config_map)
    }
}

impl ConfigMapBuilder {
    pub fn new(client: Client) -> Self {
        ConfigMapBuilder { client }
    }
}

#[cfg(test)]
mod tests {
    use shulker_crds::v1alpha1::proxy_fleet::ProxyFleetTemplateVersion;
    use shulker_kube_utils::reconcilers::builder::ResourceBuilder;

    use crate::reconcilers::proxy_fleet::fixtures::{create_client_mock, TEST_PROXY_FLEET};

    #[test]
    fn name_contains_fleet_name() {
        // W
        let name = super::ConfigMapBuilder::name(&TEST_PROXY_FLEET);

        // T
        assert_eq!(name, "my-proxy-config");
    }

    #[tokio::test]
    async fn build_snapshot() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);
        let name = super::ConfigMapBuilder::name(&TEST_PROXY_FLEET);

        // W
        let config_map = builder
            .build(&TEST_PROXY_FLEET, &name, None, None)
            .await
            .unwrap();

        // T
        insta::assert_yaml_snapshot!(config_map);
    }

    #[tokio::test]
    async fn build_has_startup_script() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);
        let name = super::ConfigMapBuilder::name(&TEST_PROXY_FLEET);

        // W
        let config_map = builder
            .build(&TEST_PROXY_FLEET, &name, None, None)
            .await
            .unwrap();

        // T
        assert!(config_map.data.as_ref().unwrap().contains_key("init-fs.sh"));
    }

    #[tokio::test]
    async fn build_has_readiness_script() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);
        let name = super::ConfigMapBuilder::name(&TEST_PROXY_FLEET);

        // W
        let config_map = builder
            .build(&TEST_PROXY_FLEET, &name, None, None)
            .await
            .unwrap();

        // T
        assert!(config_map
            .data
            .as_ref()
            .unwrap()
            .contains_key("probe-readiness.sh"));
    }

    #[tokio::test]
    async fn build_has_velocity_configs() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);
        let name = super::ConfigMapBuilder::name(&TEST_PROXY_FLEET);

        // W
        let config_map = builder
            .build(&TEST_PROXY_FLEET, &name, None, None)
            .await
            .unwrap();

        // T
        assert!(config_map
            .data
            .as_ref()
            .unwrap()
            .contains_key("server-icon.png"));
        assert!(config_map
            .data
            .as_ref()
            .unwrap()
            .contains_key("velocity-config.toml"));
    }

    #[tokio::test]
    async fn build_has_bungeecord_configs() {
        // G
        let client = create_client_mock();
        let builder = super::ConfigMapBuilder::new(client);
        let name = super::ConfigMapBuilder::name(&TEST_PROXY_FLEET);
        let mut fleet = TEST_PROXY_FLEET.clone();
        fleet.spec.template.spec.version.channel = ProxyFleetTemplateVersion::BungeeCord;

        // W
        let config_map = builder.build(&fleet, &name, None, None).await.unwrap();

        // T
        assert!(config_map
            .data
            .as_ref()
            .unwrap()
            .contains_key("server-icon.png"));
        assert!(config_map
            .data
            .as_ref()
            .unwrap()
            .contains_key("bungeecord-config.yml"));
    }
}

mod bungeecord {
    use serde::{Deserialize, Serialize};
    use std::collections::BTreeMap;

    use shulker_crds::v1alpha1::proxy_fleet::{
        ProxyFleetServiceSpec, ProxyFleetServiceType, ProxyFleetTemplateConfigurationSpec,
    };

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "snake_case")]
    pub struct BungeeCordYml {
        servers: BTreeMap<String, BungeeCordServerYml>,
        listeners: Vec<BungeeCordListenerYml>,
        groups: BTreeMap<String, ()>,
        online_mode: bool,
        ip_forward: bool,
        prevent_proxy_connections: bool,
        enforce_secure_profile: bool,
        log_pings: bool,
        reject_transfers: bool,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "snake_case")]
    pub struct BungeeCordServerYml {
        motd: String,
        address: String,
        restricted: bool,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "snake_case")]
    pub struct BungeeCordListenerYml {
        host: String,
        query_port: u16,
        motd: String,
        max_players: u32,
        priorities: Vec<String>,
        ping_passthrough: bool,
        force_default_server: bool,
        proxy_protocol: bool,
    }

    impl BungeeCordYml {
        pub fn from_spec(
            spec: &ProxyFleetTemplateConfigurationSpec,
            service_spec: Option<&ProxyFleetServiceSpec>,
        ) -> Self {
            let disallow_proxy_connections = service_spec
                .map(|spec| {
                    spec.type_ == ProxyFleetServiceType::LoadBalancer
                        || spec.type_ == ProxyFleetServiceType::NodePort
                })
                .unwrap_or(false);

            BungeeCordYml {
                servers: BTreeMap::from([
                    (
                        "lobby".to_string(),
                        BungeeCordServerYml {
                            motd: spec.motd.clone(),
                            address: "localhost:30000".to_string(),
                            restricted: false,
                        },
                    ),
                    (
                        "limbo".to_string(),
                        BungeeCordServerYml {
                            motd: spec.motd.clone(),
                            address: "localhost:30001".to_string(),
                            restricted: false,
                        },
                    ),
                ]),
                listeners: vec![BungeeCordListenerYml {
                    host: "0.0.0.0:25577".to_string(),
                    query_port: 25577,
                    motd: spec.motd.clone(),
                    max_players: spec.max_players,
                    priorities: vec!["lobby".to_string(), "limbo".to_string()],
                    ping_passthrough: false,
                    force_default_server: true,
                    proxy_protocol: spec.proxy_protocol,
                }],
                groups: BTreeMap::new(),
                online_mode: true,
                ip_forward: true,
                prevent_proxy_connections: disallow_proxy_connections,
                enforce_secure_profile: true,
                log_pings: false,
                reject_transfers: false,
            }
        }
    }

    impl ToString for BungeeCordYml {
        fn to_string(&self) -> String {
            let mut yml = serde_yaml::to_string(&self).unwrap();
            yml.push('\n');

            yml
        }
    }

    #[cfg(test)]
    mod tests {
        use std::collections::BTreeMap;

        use shulker_crds::v1alpha1::proxy_fleet::{
            ProxyFleetServiceSpec, ProxyFleetServiceType, ProxyFleetTemplateConfigurationSpec,
        };

        #[test]
        fn from_spec() {
            // G
            let spec = ProxyFleetTemplateConfigurationSpec {
                existing_config_map_name: None,
                plugins: None,
                patches: None,
                max_players: 100,
                motd: "A Motd".to_string(),
                server_icon: "A Server Icon".to_string(),
                proxy_protocol: true,
                ttl_seconds: 300,
                players_delta_before_exclusion: 15,
                skip_agent_download: false,
            };
            let service_spec = Some(ProxyFleetServiceSpec {
                type_: ProxyFleetServiceType::LoadBalancer,
                ..ProxyFleetServiceSpec::default()
            });

            // W
            let config = super::BungeeCordYml::from_spec(&spec, service_spec.as_ref());

            // T
            insta::assert_yaml_snapshot!(config);
        }

        #[test]
        fn from_spec_prevent_proxy_connections_when_nodeport() {
            // G
            let spec = ProxyFleetTemplateConfigurationSpec::default();
            let service_spec = Some(ProxyFleetServiceSpec {
                type_: ProxyFleetServiceType::NodePort,
                ..ProxyFleetServiceSpec::default()
            });

            // W
            let config = super::BungeeCordYml::from_spec(&spec, service_spec.as_ref());

            // T
            assert!(config.prevent_proxy_connections);
        }

        #[test]
        fn from_spec_not_prevent_proxy_connections_when_clusterip() {
            // G
            let spec = ProxyFleetTemplateConfigurationSpec::default();
            let service_spec = Some(ProxyFleetServiceSpec {
                type_: ProxyFleetServiceType::ClusterIP,
                ..ProxyFleetServiceSpec::default()
            });

            // W
            let config = super::BungeeCordYml::from_spec(&spec, service_spec.as_ref());

            // T
            assert!(!config.prevent_proxy_connections);
        }

        #[test]
        fn from_spec_not_prevent_proxy_connections_when_no_service() {
            // G
            let spec = ProxyFleetTemplateConfigurationSpec::default();
            let service_spec = None;

            // W
            let config = super::BungeeCordYml::from_spec(&spec, service_spec.as_ref());

            // T
            assert!(!config.prevent_proxy_connections);
        }

        #[test]
        fn to_string() {
            // G
            let config = super::BungeeCordYml {
                servers: BTreeMap::from([(
                    "lobby".to_string(),
                    super::BungeeCordServerYml {
                        motd: "Hello World".to_string(),
                        address: "localhost:30000".to_string(),
                        restricted: false,
                    },
                )]),
                listeners: vec![super::BungeeCordListenerYml {
                    host: "0.0.0.0:25577".to_string(),
                    query_port: 25577,
                    motd: "Hello World".to_string(),
                    max_players: 20,
                    priorities: vec!["lobby".to_string(), "limbo".to_string()],
                    ping_passthrough: false,
                    force_default_server: true,
                    proxy_protocol: true,
                }],
                groups: BTreeMap::new(),
                online_mode: true,
                ip_forward: true,
                prevent_proxy_connections: true,
                enforce_secure_profile: true,
                log_pings: false,
                reject_transfers: false,
            };

            // W
            let yml = config.to_string();

            // T
            insta::assert_snapshot!(yml);
        }
    }
}

mod velocity {
    use serde::{Deserialize, Serialize};

    use shulker_crds::v1alpha1::proxy_fleet::{
        ProxyFleetServiceSpec, ProxyFleetServiceType, ProxyFleetTemplateConfigurationSpec,
    };

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct VelocityToml {
        config_version: String,
        bind: String,
        motd: String,
        show_max_players: u32,
        online_mode: bool,
        force_key_authentication: bool,
        prevent_client_proxy_connections: bool,
        forwarding_secret_file: String,
        player_info_forwarding_mode: String,
        servers: VelocityServersToml,
        forced_hosts: VelocityForcedHostsToml,
        advanced: VelocityAdvancedToml,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct VelocityServersToml {
        lobby: String,
        limbo: String,
        #[serde(rename = "try")]
        try_: Vec<String>,
    }

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct VelocityForcedHostsToml {}

    #[derive(Deserialize, Serialize, Clone, Debug)]
    #[serde(rename_all = "kebab-case")]
    pub struct VelocityAdvancedToml {
        haproxy_protocol: bool,
        tcp_fast_open: bool,
        accepts_transfers: bool,
    }

    impl VelocityToml {
        pub fn from_spec(
            spec: &ProxyFleetTemplateConfigurationSpec,
            service_spec: Option<&ProxyFleetServiceSpec>,
        ) -> Self {
            let disallow_proxy_connections = service_spec
                .map(|spec| {
                    spec.type_ == ProxyFleetServiceType::LoadBalancer
                        || spec.type_ == ProxyFleetServiceType::NodePort
                })
                .unwrap_or(false);

            VelocityToml {
                config_version: "2.7".to_string(),
                bind: "0.0.0.0:25577".to_string(),
                motd: spec.motd.clone(),
                show_max_players: spec.max_players,
                online_mode: true,
                force_key_authentication: true,
                prevent_client_proxy_connections: disallow_proxy_connections,
                player_info_forwarding_mode: "modern".to_string(),
                forwarding_secret_file: "/mnt/shulker/forwarding-secret/key".to_string(),
                servers: VelocityServersToml {
                    lobby: "localhost:30000".to_string(),
                    limbo: "localhost:30001".to_string(),
                    try_: vec!["lobby".to_string(), "limbo".to_string()],
                },
                forced_hosts: VelocityForcedHostsToml {},
                advanced: VelocityAdvancedToml {
                    haproxy_protocol: spec.proxy_protocol,
                    tcp_fast_open: true,
                    accepts_transfers: true,
                },
            }
        }
    }

    impl ToString for VelocityToml {
        fn to_string(&self) -> String {
            let mut toml = toml::to_string(&self).unwrap();
            toml.push('\n');

            toml
        }
    }

    #[cfg(test)]
    mod tests {
        use shulker_crds::v1alpha1::proxy_fleet::{
            ProxyFleetServiceSpec, ProxyFleetServiceType, ProxyFleetTemplateConfigurationSpec,
        };

        #[test]
        fn from_spec() {
            // G
            let spec = ProxyFleetTemplateConfigurationSpec {
                existing_config_map_name: None,
                plugins: None,
                patches: None,
                max_players: 100,
                motd: "A Motd".to_string(),
                server_icon: "A Server Icon".to_string(),
                proxy_protocol: true,
                ttl_seconds: 300,
                players_delta_before_exclusion: 15,
                skip_agent_download: false,
            };
            let service_spec = Some(ProxyFleetServiceSpec {
                type_: ProxyFleetServiceType::LoadBalancer,
                ..ProxyFleetServiceSpec::default()
            });

            // W
            let config = super::VelocityToml::from_spec(&spec, service_spec.as_ref());

            // T
            insta::assert_toml_snapshot!(config);
        }

        #[test]
        fn from_spec_prevent_proxy_connections_when_nodeport() {
            // G
            let spec = ProxyFleetTemplateConfigurationSpec::default();
            let service_spec = Some(ProxyFleetServiceSpec {
                type_: ProxyFleetServiceType::NodePort,
                ..ProxyFleetServiceSpec::default()
            });

            // W
            let config = super::VelocityToml::from_spec(&spec, service_spec.as_ref());

            // T
            assert!(config.prevent_client_proxy_connections);
        }

        #[test]
        fn from_spec_not_prevent_proxy_connections_when_clusterip() {
            // G
            let spec = ProxyFleetTemplateConfigurationSpec::default();
            let service_spec = Some(ProxyFleetServiceSpec {
                type_: ProxyFleetServiceType::ClusterIP,
                ..ProxyFleetServiceSpec::default()
            });

            // W
            let config = super::VelocityToml::from_spec(&spec, service_spec.as_ref());

            // T
            assert!(!config.prevent_client_proxy_connections);
        }

        #[test]
        fn from_spec_not_prevent_proxy_connections_when_no_service() {
            // G
            let spec = ProxyFleetTemplateConfigurationSpec::default();
            let service_spec = None;

            // W
            let config = super::VelocityToml::from_spec(&spec, service_spec.as_ref());

            // T
            assert!(!config.prevent_client_proxy_connections);
        }

        #[test]
        fn to_string() {
            // G
            let config = super::VelocityToml {
                config_version: "2.6".to_string(),
                bind: "0.0.0.0:25577".to_string(),
                motd: "A Motd".to_string(),
                show_max_players: 100,
                online_mode: true,
                force_key_authentication: true,
                prevent_client_proxy_connections: true,
                forwarding_secret_file: "forwarding-secret.txt".to_string(),
                player_info_forwarding_mode: "modern".to_string(),
                servers: super::VelocityServersToml {
                    lobby: "localhost:30000".to_string(),
                    limbo: "localhost:30001".to_string(),
                    try_: vec!["lobby".to_string(), "limbo".to_string()],
                },
                forced_hosts: super::VelocityForcedHostsToml {},
                advanced: super::VelocityAdvancedToml {
                    haproxy_protocol: true,
                    tcp_fast_open: true,
                    accepts_transfers: true,
                },
            };

            // W
            let toml = config.to_string();

            // T
            insta::assert_snapshot!(toml);
        }
    }
}
