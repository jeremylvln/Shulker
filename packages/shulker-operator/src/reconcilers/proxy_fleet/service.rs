use k8s_openapi::api::core::v1::Service;
use k8s_openapi::api::core::v1::ServicePort;
use k8s_openapi::api::core::v1::ServiceSpec;
use k8s_openapi::apimachinery::pkg::util::intstr::IntOrString;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use shulker_crds::v1alpha1::proxy_fleet::ProxyFleet;

use super::ProxyFleetReconciler;

pub struct ServiceBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for ServiceBuilder {
    type OwnerType = ProxyFleet;
    type ResourceType = Service;

    fn name(proxy_fleet: &Self::OwnerType) -> String {
        proxy_fleet.name_any()
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, proxy_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            proxy_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, proxy_fleet: &Self::OwnerType) -> bool {
        proxy_fleet.spec.service.is_some()
    }

    async fn create(
        &self,
        proxy_fleet: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let service = Service {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(proxy_fleet.namespace().unwrap().clone()),
                labels: Some(
                    ProxyFleetReconciler::get_common_labels(proxy_fleet)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            ..Service::default()
        };

        Ok(service)
    }

    async fn update(
        &self,
        proxy_fleet: &Self::OwnerType,
        service: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        let service_config = proxy_fleet.spec.service.as_ref().unwrap();

        service.spec = Some(ServiceSpec {
            selector: Some(
                ProxyFleetReconciler::get_common_labels(proxy_fleet)
                    .into_iter()
                    .collect(),
            ),
            type_: Some(service_config.type_.to_string()),
            external_traffic_policy: service_config
                .external_traffic_policy
                .as_ref()
                .map(|x| x.to_string()),
            ports: Some(vec![ServicePort {
                name: Some("minecraft".to_string()),
                protocol: Some("TCP".to_string()),
                port: 25565,
                target_port: Some(IntOrString::Int(25577)),
                ..ServicePort::default()
            }]),
            ..ServiceSpec::default()
        });

        Ok(())
    }
}

impl ServiceBuilder {
    pub fn new(client: Client) -> Self {
        ServiceBuilder { client }
    }
}
