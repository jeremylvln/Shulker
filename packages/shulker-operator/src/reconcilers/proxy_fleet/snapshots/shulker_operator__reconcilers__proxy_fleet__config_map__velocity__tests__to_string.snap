---
source: packages/shulker-operator/src/reconcilers/proxy_fleet/config_map.rs
expression: toml
---
config-version = "2.6"
bind = "0.0.0.0:25577"
motd = "A Motd"
show-max-players = 100
online-mode = true
force-key-authentication = true
prevent-client-proxy-connections = true
forwarding-secret-file = "forwarding-secret.txt"
player-info-forwarding-mode = "modern"

[servers]
lobby = "localhost:30000"
limbo = "localhost:30001"
try = ["lobby", "limbo"]

[forced-hosts]

[advanced]
haproxy-protocol = true
tcp-fast-open = true
accepts-transfers = true
